### DISCUSSION TITLE
How can a subclass module add a new io to the original io of the parent class module?(请问子类module如何在父类module原有的io基础上增加新的io？)

### DISCUSSION THREAD

**Discussion by ddxoxpp** [chipsalliance/chisel#2227] [Q&A]
As the title.
(如题。)

**Comment by sequencer** [comment]
TL;DR
There are three methods:
1. [Aspect](https://github.com/chipsalliance/chisel3/blob/master/core/src/main/scala/chisel3/aop/Aspect.scala) API
2. [InModuleBody](https://github.com/chipsalliance/diplomacy/blob/master/diplomacy/src/diplomacy/lazymodule/InModuleBody.scala) API from [diplomacy](https://github.com/chipsalliance/diplomacy/blob/master/diplomacy)
3. [registerFinishAction](https://github.com/chipsalliance/chisel3/pull/1943) API(unmerged yet)

when involving chisel internal implementations, for a life cycle of a Module, there are three periods:
1. Before Module instantiated:
you are free to register any code blocks to a list buffer, and instantiate it later when Module is instantiating.
2. When Module is instantiating, there are two different cases:
- inside *this* module, `Module.currentModule` is this module: you can add any `IO` or something else to your Module, like what we usually does.
- outside of *this* module, generally this is a case, when Module stack goes into another module, in this case, you are not allowed to add code blocks to the *this* module, Diplomacy resolves this via lazy implementation via this [code block](https://github.com/chipsalliance/diplomacy/blob/252005dd00e471fcbc6377584d3f5e48b85ba87e/diplomacy/src/diplomacy/lazymodule/LazyModuleImp.scala#L89-L92), if #1943 merged, you can use that API to resolve this as well.
3. After Module instantiated,
you are forbid to add any IR to *this* module, thus you can only use Aspect API to match the module and add code block later(the API is fairly complicated, @azidar may help)

So let's go back to your question:
> How can a subclass module add a new io to the original io of the parent class module?

If you are not willing to use diplomacy currently, the simplest API might be 3, but not merged yet. We can discuss this question in today's dev-meeting.

**Reply by ekiwi** [reply]
Why can't you just do:

```.scala
class Parent extends Module {
  val parentIn = IO(Input(Bool()))
  ....
}

class Child extends Parent {
  val additionalIn = IO(Input(Bool()))
  ....
}
```

**Reply by edwardcwang** [reply]
How would someone do it if the parent module uses a traditional IO bundle?

**Reply by ekiwi** [reply]
So is the question about _overwriting_ a parent I/O? If so, I am not sure if any of the other proposed solutions would work.
In that case it might make sense to not have any I/Os in the parent class and just to make two different child classes each with different I/O.

**Reply by edwardcwang** [reply]
Because the traditional Chisel I/O style is

```scala
class Parent extends Module {
  val io = IO(new Bundle {
    val foo = Input(Bool())
    val bar = Output(UInt(10.W))
  }
  // ...
}
}

**Reply by ekiwi** [reply]
Yeah, this is a situation where the best answer may be to just use two separate I/Os instead of a single `io` bundle.

**Reply by sequencer** [reply]
My bad, I misunderstood the original question, I thought it was modifying the outer module from inner module.

**Reply by ddxoxpp** [reply]
I am sorry for not clarifying the situation:
```
class Parent extends BoomModule {
  val io = IO(new Bundle {
    ...
  }
 class Child extends Parent {
  val add_io = IO(new Bundle {
    ...
  }
```
And then make failed.

**Reply by jackkoenig** [reply]
What is the error?

**Comment by mwachs5** [comment]
what is a `BoomModule`? In order for what you are doing to work, it needs to be a `chisel3.MultiIOModule` or a `chisel3.Module` (3.5 only) not a `chisel3.Module` (3.4.x or earlier) or a `Chisel.Module`.

**Reply by ddxoxpp** [reply]
```BoomModule``` is used in ```chipyard/generators/boom```, it seems to be a ```MultiIOModule```
